### PR TITLE
feat: support edfs subscription stream/consumer; multiple subjects

### PR DESCRIPTION
**Pull request by Aenimus**
## TODO

- [ ] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

File: router/pkg/pubsub/nats.go [hunk 1]
```diff
@@ -43,46 +57,87 @@ func (p *natsPubSub) ID() string {
...
+					updater.Update(msg.Data())
+				}
 			}
-			updater.Update(msg.Data)
+		}()
+		return nil
+	}
+
+	for _, subject := range subjects {
+		subscription, err := p.conn.SubscribeSync(subject)
```
Comment (jensneuse) [inline router/pkg/pubsub/nats.go:109]: I don't think this is the right approach.
I'd use ChanSubscribe multiple times with the same channel as the "sink".
Then use one single goroutine to read on this channel and emit to updater.
Let me know if you'd like to discuss this.

**Review by jensneuse — CHANGES_REQUESTED** [review]

File: router/gen/proto/wg/cosmo/node/v1/node.pb.go [hunk 2]
```diff
@@ -1850,22 +1850,78 @@ func (x *DataSourceCustom_GraphQL) GetCustomScalarTypeFields() []*SingleTypeFiel
...
+
+func (x *StreamConfiguration) GetConsumer() string {
+	if x != nil {
+		return x.Consumer
+	}
+	return ""
+}
+
+func (x *StreamConfiguration) GetSteamName() string {
+	if x != nil {
```
Comment (StarpTech) [inline router/gen/proto/wg/cosmo/node/v1/node.pb.go:1902]: Typo

File: proto/wg/cosmo/node/v1/node.proto [hunk 3]
```diff
@@ -199,12 +199,18 @@ enum EventType {
   SUBSCRIBE = 2;
 }
 
+message StreamConfiguration {
+  string consumer = 1;
+  string steam_name = 2;
```
Comment (StarpTech) [inline proto/wg/cosmo/node/v1/node.proto:204]: Typo

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: proto/wg/cosmo/node/v1/node.proto (hunk 3 above)
Comment (Aenimus) [reply proto/wg/cosmo/node/v1/node.proto:204]: 🙈

File: router/pkg/pubsub/nats.go (hunk 1 above)
Comment (Aenimus) [reply router/pkg/pubsub/nats.go]: Changed.

File: demo/pkg/subgraphs/employeeupdated/subgraph/schema.graphqls [hunk 4]
```diff
@@ -17,15 +13,25 @@ input UpdateEmployeeInput {
...
+    updateEmployee(id: ID!, update: UpdateEmployeeInput!): edfs__PublishEventResult! @edfs__eventsPublish(subject: "updateEmployee.{{ args.id }}")
+    updateEmployeeMyNats(employeeID: ID!, update: UpdateEmployeeInput!): edfs__PublishEventResult! @edfs__eventsPublish(subject: "updateEmployeeMyNats.{{ args.employeeID }}", sourceName: "my-nats")
 }
 
 type Subscription {
-    employeeUpdated(employeeID: ID!): Employee! @eventsSubscribe(topic: "employeeUpdated.{{ args.employeeID }}")
-    employeeUpdatedMyNats(id: ID!): Employee! @eventsSubscribe(topic: "employeeUpdatedMyNats.{{ args.id }}", sourceName: "my-nats")
+    employeeUpdated(employeeID: ID!): Employee! @edfs__eventsSubscribe(subjects: ["employeeUpdated.{{ args.employeeID }}"])
+    employeeUpdatedMyNats(id: ID!): Employee! @edfs__eventsSubscribe(subjects: ["employeeUpdatedMyNats.{{ args.id }}", "employeeUpdatedMyNatsTwo.{{ args.id }}"], sourceName: "my-nats")
+    employeeUpdatedStream(id: ID!): Employee! @edfs__eventsSubscribe(subjects: ["employeeUpdated.{{ args.id }}"], streamConfiguration: { consumer: "consumerName", streamName: "streamName"})
```
[the remark this answers is not in the document]
Comment (Aenimus) [reply demo/pkg/subgraphs/employeeupdated/subgraph/schema.graphqls:23]: I added the creation of a stream in test1 subgraph.go.

File: router/pkg/pubsub/nats.go [hunk 5]
```diff
@@ -43,46 +57,93 @@ func (p *natsPubSub) ID() string {
...
+	}
+
+	msgChan := make(chan *nats.Msg)
+	subscriptions := make([]*nats.Subscription, len(subjects))
+	for _, subject := range subjects {
+		subscription, err := p.conn.ChanSubscribe(subject, msgChan)
+		if err != nil {
+			return newEDFSNatsError(fmt.Errorf(`error subscribing to NATS subject "%s": %w`, subject, err))
+		}
+		subscriptions = append(subscriptions, subscription)
```
Comment (jensneuse) [inline router/pkg/pubsub/nats.go:115]: It's a nitpick, but you could also add and index "i" in line 110 and assign using the index, no need to append.

File: demo/pkg/subgraphs/subgraphs.go [hunk 6]
```diff
@@ -180,6 +180,20 @@ func New(ctx context.Context, config *Config) (*Subgraphs, error) {
 		"default": pubsub.NewNATSConnector(defaultConnection).New(ctx),
 		"my-nats": pubsub.NewNATSConnector(myNatsConnection).New(ctx),
 	}
+
+	js, err := jetstream.New(defaultConnection)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = js.CreateStream(ctx, jetstream.StreamConfig{
```
Comment (jensneuse) [inline demo/pkg/subgraphs/subgraphs.go:189]: Is this idempotent or will it fail if the stream already exists?

**Review by jensneuse — CHANGES_REQUESTED** [review]

File: router/pkg/pubsub/nats.go (hunk 5 above)
Comment (Aenimus) [reply router/pkg/pubsub/nats.go:115]: Got it

File: demo/pkg/subgraphs/subgraphs.go (hunk 6 above)
Comment (Aenimus) [reply demo/pkg/subgraphs/subgraphs.go:189]: I have changed to `CreateOrUpdateStream`

**Review by jensneuse — APPROVED** [review]
LGTM

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**